kernel/kernel-api/src/main/java/io/delta/kernel/internal/IcebergCompatV2Utils.java

```diff
@@ -33,4 +66,216 @@ public static void validDataFileStatus(DataFileStatus dataFileStatus) {
+          // with the IcebergV1 requirements, we fail
+          throw new UnsupportedOperationException(
+              String.format(
+                  "The value '%s' for the property '%s' is not compatible with "
+                      + "IcebergV2 requirements",
+                  newestValue, propertyName));
+        }
+      }
+
+      return Optional.empty();
+    }
+  }
+
+  private abstract static class IcebergCompatCheck {
+    abstract void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol);
+  }
+
+  private static final RequiredDeltaTablePropertyEnforcer ICEBERG_COMPAT_V2_CM_REQUIREMENT =
+      new RequiredDeltaTablePropertyEnforcer(
+          TableConfig.COLUMN_MAPPING_MODE.getKey(),
+          (value) ->
+              value.equals(ColumnMappingMode.NAME.name())
+                  || value.equals(ColumnMappingMode.ID.name()),
+          ColumnMappingMode.NAME.value,
+          new PostMetadataProcessor() {
+            @Override
+            Optional<Metadata> postProcess(Metadata newMetadata, boolean isCreatingNewTable) {
+              // TODO: assign column mapping field ids
+              return Optional.empty();
+            }
+          });
+
+  private static final IcebergCompatCheck ICEBERG_COMPAT_V2_CHECK_NO_COMPAT_V1_ENABLED =
+      new IcebergCompatCheck() {
+        @Override
+        void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+          if (Boolean.valueOf(
+              newMetadata
+                  .getConfiguration()
+                  .getOrDefault("delta.enableIcebergCompatV1", "false"))) {
+            throw new IllegalArgumentException(
+                "Not expected to have delta.enableIcebergCompatV1 enabled");
+          }
+        }
+      };
+
+  private static final IcebergCompatCheck ICEBERG_COMPAT_V2_CHECK_HAS_SUPPORTED_TYPES =
+      new IcebergCompatCheck() {
+        @Override
+        void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+          // TODO
+        }
+      };
+
+  private static final IcebergCompatCheck ICEBERG_COMPAT_V2_CHECK_HAS_ALLOWED_PARTITION_TYPES =
+      new IcebergCompatCheck() {
+        @Override
+        void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+          // TODO
+        }
+      };
+
+  private static final IcebergCompatCheck ICEBERG_COMPAT_V2_CHECK_HAS_NO_PARTITION_EVOLUTION =
+      new IcebergCompatCheck() {
+        @Override
+        void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+          // TODO
+        }
+      };
+
+  private static final IcebergCompatCheck ICEBERG_COMPAT_V2_CHECK_HAS_NO_DELETION_VECTORS =
+      new IcebergCompatCheck() {
+        @Override
+        void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+          // TODO
+        }
+      };
+
+  private static class IcebergCompatMetadataValidatorAndUpdater {
```

Comment:
so this is our version of `IcebergCompat` right?

Reply:
I think it might help just organization wise to put all the generic interfaces in one file, and then for each iceberg compat table feature define the implementations in their own file. But maybe we can also just group these elements differently not sure.

But I think `IcebergCompatMetadataValidatorAndUpdater` seems fine to be it's own top-level class

Reply:
separated the validator and updated into a separate class

Reply:
Did you mean to remove this code that is now in the separate file? Sorry left comments here